### ISSUE TITLE
SegmentStore does not respond to READ_SEGMENT for empty segments when the segment is sealed. 

### ISSUE THREAD

**Issue by shiveshr** [pravega/pravega#2679]
**Problem description**
The ReadWithAutoScaleTest system test failed with readers not having read all the written events. 
Readers received the scaled up segments post couple of scales, and from this point, they continued returning NULL events every 1 minute.
These segments are empty and read future is waiting in the background when they were sealed. However, client never received `SegmentSealed` response and thus continued waiting on READ_SEGMENT call to respond.

Following is the flow of epoch transitions because of scale and rolling txns. 
```
1. epoch 0: 0.0, 1.0
2. scale to epoch 1: 0.0, 2.1, 3.1
3. scale to epoch 2: 2.1, 3.1, 4.2, 5.2
4. rollingTxn to epoh 3: 0.3, 1.3  // duplicate of epoch 0
    rollingTxn to epoch 4: 2.4, 3.4, 5.4 // duplicate of epoch 2
5. rollingTxn to epoch 5: 0.5, 1.5 // duplicate of epoch 0
    rollingTxn to epoch 6: 2.6, 3.6, 4.6 // duplicate of epoch 2
```

```
// readers receive segmentSealed for 0.0 and 1.0 and request their successors
15:13:44,035 130896 [ForkJoinPool.commonPool-worker-2] TRACE i.p.c.stream.impl.ControllerImpl - ENTER getSuccessors@11229449082272 [testAutoScale1765265652/testTxnScaleUpWithRead/1#epoch.0].
15:13:44,051 // successors received = 4294967298 (2.1) and 4294967299 (3.1)

15:13:46,564 133425 [ForkJoinPool.commonPool-worker-3] TRACE i.p.c.stream.impl.ControllerImpl - ENTER getSuccessors@11231978213085 [testAutoScale1765265652/testTxnScaleUpWithRead/0#epoch.0].
15:13:46,573 // successors received 8589934596 (4.2) and 8589934597 (5.2)

// Readers continue to return NULL event from this point on. All four segments, 2.1 and 3.1 and 4.2 and 5.2 are empty. 
// Also they are sealed soon enough because of rollingTxn. But readers receive no notification of the same and continue waiting on READ_SEGMENT's response from segment store. 

// reader group state
15:15:40,274 247135 [ForkJoinPool.commonPool-worker-2] DEBUG i.p.c.s.i.ReaderGroupStateManager - Created new checkpoint: io.pravega.client.stream.impl.ReaderGroupState$CreateCheckpoint@1 currentState is: ReaderGroupState{ CheckpointState { ongoingCheckpoints: [8fa3de49-0b2e-4ba8-97e4-2de0d49e3b32],  readersBlockingEachCheckpoint: {} } futureSegments: {} assignedSegments: {reader2={testAutoScale1765265652/testTxnScaleUpWithRead/4#epoch.2=0, testAutoScale1765265652/testTxnScaleUpWithRead/5#epoch.2=0}, reader1={testAutoScale1765265652/testTxnScaleUpWithRead/3#epoch.1=0, testAutoScale1765265652/testTxnScaleUpWithRead/2#epoch.1=0}} unassignedSegments: {} }

// SEGMENT_READ issued for all 4 segments, (printing only one of them here)
15:13:46,647 7359784 [epollEventLoopGroup-3-6] DEBUG i.p.s.s.h.h.ServerConnectionInboundHandler - Processing request: WireCommands.ReadSegment(type=READ_SEGMENT, segment=testAutoScale1765265652/testTxnScaleUpWithRead/5#epoch.2, offset=0, suggestedLength=65536, delegationToken=)
Since these segments are emtpty, so there is no SEGMENT_READ response for any of these 4 requests. 
However, 10 seconds later these segments are seal. Yes, segment store does not respond with SEGMENT_SEALED and readers continue waiting. 

// The test finds scale to have happened and decides to complete it by first stopping the writer and then waiting on readers to read all events
15:13:54,761 141622 [ForkJoinPool.commonPool-worker-0] INFO  i.p.t.system.ReadWithAutoScaleTest - All writers have stopped. Setting Stop_Read_Flag. Event Written Count:21900, Event Read Count: 20300

15:14:41,192 188053 [ForkJoinPool.commonPool-worker-2] DEBUG i.p.t.system.ReadWithAutoScaleTest - Null event, reader reader1, read count 20300
... // NULL event repeats every one minute for both readers

15:22:41,033 667894 [ForkJoinPool.commonPool-worker-2] DEBUG i.p.t.system.ReadWithAutoScaleTest - Null event, reader reader1, read count 20300
```

these segments were sealed at around 15:13:52
```
15:13:52,531 7504380 [epollEventLoopGroup-2-5] DEBUG i.p.c.n.i.ClientConnectionInboundHandler - 172.16.18.25 processing reply: WireCommands.SegmentSealed(type=SEGMENT_SEALED, requestId=11979, segment=testAutoScale1765265652/testTxnScaleUpWithRead/2#epoch.1)

15:13:52,530 7504379 [epollEventLoopGroup-2-4] DEBUG i.p.c.n.i.ClientConnectionInboundHandler - 172.16.18.25 processing reply: WireCommands.SegmentSealed(type=SEGMENT_SEALED, requestId=11978, segment=testAutoScale1765265652/testTxnScaleUpWithRead/3#epoch.1)

15:13:52,531 7504380 [epollEventLoopGroup-2-2] DEBUG i.p.c.n.i.ClientConnectionInboundHandler - 172.16.18.25 processing reply: WireCommands.SegmentSealed(type=SEGMENT_SEALED, requestId=11976, segment=testAutoScale1765265652/testTxnScaleUpWithRead/4#epoch.2)

15:13:52,532 7504381 [epollEventLoopGroup-2-3] DEBUG i.p.c.n.i.ClientConnectionInboundHandler - 172.16.18.25 processing reply: WireCommands.SegmentSealed(type=SEGMENT_SEALED, requestId=11977, segment=testAutoScale1765265652/testTxnScaleUpWithRead/5#epoch.2)
```

I am not sure if these are interesting at all. 
```
2018-06-08 15:13:58,430 7371567 [core-24] WARN  i.p.s.s.a.SegmentAttributeIndex - AttributeIndex[67647]: Snapshot serialized to attribute segment, but failed to update snapshot location due to io.pravega.segmentstore.contracts.StreamSegmentSealedException: [Segment 'testAutoScale1765265652/testTxnScaleUpWithRead/2#epoch.1'] The StreamSegment is sealed and cannot be modified..
2018-06-08 15:13:58,440 7371577 [core-16] WARN  i.p.s.s.a.SegmentAttributeIndex - AttributeIndex[67671]: Snapshot serialized to attribute segment, but failed to update snapshot location due to io.pravega.segmentstore.contracts.StreamSegmentSealedException: [Segment 'testAutoScale1765265652/testTxnScaleUpWithRead/4#epoch.2'] The StreamSegment is sealed and cannot be modified..
2018-06-08 15:13:59,185 7372322 [core-17] WARN  i.p.s.s.a.SegmentAttributeIndex - AttributeIndex[67646]: Snapshot serialized to attribute segment, but failed to update snapshot location due to io.pravega.segmentstore.contracts.StreamSegmentSealedException: [Segment 'testAutoScale1765265652/testTxnScaleUpWithRead/3#epoch.1'] The StreamSegment is sealed and cannot be modified..
```

**Problem location**
Investigate more.
My initial suspect is server side because reader never received SegmentSealedException. 

**Suggestions for an improvement**
Fix

**Comment by shiveshr** [comment]
Following is the flow of epoch transitions because of scale and rolling txns. 
```
1. epoch 0: 0.0, 1.0
2. scale to epoch 1: 0.0, 2.1, 3.1
3. scale to epoch 2: 2.1, 3.1, 4.2, 5.2
4. rollingTxn to epoh 3: 0.3, 1.3  // duplicate of epoch 0
    rollingTxn to epoch 4: 2.4, 3.4, 5.4 // duplicate of epoch 2
5. rollingTxn to epoch 5: 0.5, 1.5 // duplicate of epoch 0
    rollingTxn to epoch 6: 2.6, 3.6, 4.6 // duplicate of epoch 2
```

- readers found segments 0.0 and 1.0 to be sealed and duly requested for their successors at 15:13:44 and 15:13:46 
So the set of segments readers are working with when they stop reading are `2.1, 3.1, 4.2, 5.2`

All of the above segments were sealed at 15:13:52 and client never read any data from these segment, nor did it receive SegmentSealed exception. 
We notice first NULL event for reader1 and reader2 at 15:14:41 which is returned after 1 minute of lack of data.. 
so this implies reader1 and reader2 never read anything after 15:13:41

**Comment by shiveshr** [comment]
another reproduction of the issue with client side logs
[1176_client.zip](https://github.com/pravega/pravega/files/2087522/1176_client.zip)

**Comment by shiveshr** [comment]
i can reproduce the issue locally with following code
```
/**
 * Copyright (c) 2017 Dell Inc., or its subsidiaries. All Rights Reserved.
 * <p>
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p>
 * http://www.apache.org/licenses/LICENSE-2.0
 */
package io.pravega.test.integration.demo;

import io.pravega.client.ClientConfig;
import io.pravega.client.ClientFactory;
import io.pravega.client.admin.ReaderGroupManager;
import io.pravega.client.admin.impl.ReaderGroupManagerImpl;
import io.pravega.client.netty.impl.ConnectionFactory;
import io.pravega.client.netty.impl.ConnectionFactoryImpl;
import io.pravega.client.stream.EventRead;
import io.pravega.client.stream.EventStreamReader;
import io.pravega.client.stream.EventStreamWriter;
import io.pravega.client.stream.EventWriterConfig;
import io.pravega.client.stream.ReaderConfig;
import io.pravega.client.stream.ReaderGroupConfig;
import io.pravega.client.stream.ReinitializationRequiredException;
import io.pravega.client.stream.ScalingPolicy;
import io.pravega.client.stream.Stream;
import io.pravega.client.stream.StreamConfiguration;
import io.pravega.client.stream.Transaction;
import io.pravega.client.stream.impl.ClientFactoryImpl;
import io.pravega.client.stream.impl.Controller;
import io.pravega.client.stream.impl.JavaSerializer;
import io.pravega.client.stream.impl.StreamImpl;
import io.pravega.client.stream.mock.MockClientFactory;
import io.pravega.controller.util.Config;
import io.pravega.segmentstore.contracts.StreamSegmentStore;
import io.pravega.segmentstore.server.host.handler.PravegaConnectionListener;
import io.pravega.segmentstore.server.host.stat.AutoScalerConfig;
import io.pravega.segmentstore.server.host.stat.SegmentStatsFactory;
import io.pravega.segmentstore.server.host.stat.SegmentStatsRecorder;
import io.pravega.segmentstore.server.store.ServiceBuilder;
import io.pravega.segmentstore.server.store.ServiceBuilderConfig;
import io.pravega.shared.NameUtils;
import io.pravega.test.common.TestingServerStarter;
import lombok.extern.slf4j.Slf4j;
import org.apache.curator.test.TestingServer;

import java.util.Arrays;
import java.util.Collections;
import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;

import static java.util.concurrent.TimeUnit.SECONDS;

@Slf4j
public class EndToEndAutoScaleUpWithTxnTest {
    static final StreamConfiguration CONFIG =
            StreamConfiguration.builder().scope("test").streamName("test").scalingPolicy(
                    ScalingPolicy.byEventRate(10, 2, 3)).build();

    public static void main(String[] args) throws Exception {
        TestingServer zkTestServer = new TestingServerStarter().start();
        int port = Config.SERVICE_PORT;

        ControllerWrapper controllerWrapper = new ControllerWrapper(zkTestServer.getConnectString(), port);
        Controller controller = controllerWrapper.getController();
        controllerWrapper.getControllerService().createScope(NameUtils.INTERNAL_SCOPE_NAME).get();


        ConnectionFactory connectionFactory = new ConnectionFactoryImpl(ClientConfig.builder().build());

        ClientFactory internalCF = new ClientFactoryImpl(NameUtils.INTERNAL_SCOPE_NAME, controller, connectionFactory);

        ServiceBuilder serviceBuilder = ServiceBuilder.newInMemoryBuilder(ServiceBuilderConfig.getDefaultConfig());
        serviceBuilder.initialize();
        StreamSegmentStore store = serviceBuilder.createStreamSegmentService();

        SegmentStatsFactory segmentStatsFactory = new SegmentStatsFactory();
        SegmentStatsRecorder statsRecorder = segmentStatsFactory.createSegmentStatsRecorder(store,
                internalCF,
                AutoScalerConfig.builder().with(AutoScalerConfig.MUTE_IN_SECONDS, 0)
                        .with(AutoScalerConfig.COOLDOWN_IN_SECONDS, 0).build());


        PravegaConnectionListener server = new PravegaConnectionListener(false, "localhost", 12345, store,
                statsRecorder, null, null, null);
        server.startListening();

        controllerWrapper.awaitRunning();
        controllerWrapper.getControllerService().createScope("test").get();

        controller.createStream(CONFIG).get();
        System.err.println(" stream created");
        MockClientFactory clientFactory = new MockClientFactory("test", controller);

        // Mocking pravega service by putting scale up and scale down requests for the stream
        EventWriterConfig writerConfig = EventWriterConfig.builder()
                .transactionTimeoutTime(30000)
                .build();
        EventStreamWriter<String> test = clientFactory.createEventWriter("test", new JavaSerializer<>(), writerConfig);

        ReaderGroupManager readerGroupManager = new ReaderGroupManagerImpl("test", controller, clientFactory, connectionFactory);
        readerGroupManager.createReaderGroup("readergrp", ReaderGroupConfig.builder().stream("test/test").build());

        final EventStreamReader<String> reader = clientFactory.createReader("1",
                "readergrp",
                new JavaSerializer<>(),
                ReaderConfig.builder().build());

        CompletableFuture<Void> waitForIt = new CompletableFuture<>();

        CompletableFuture.runAsync(() -> {
            try {
                while (true) {
                    EventRead<String> event = reader.readNextEvent(SECONDS.toMillis(60));
                    System.err.println("event read = " + event.getEvent());
                    if ("2".equals(event.getEvent())) {
                        waitForIt.complete(null);
                    }
                }
            } catch (ReinitializationRequiredException e) {
                e.printStackTrace();
            }
        });
        System.err.println("reader created and started in background");

        // region Successful commit tests
        Transaction<String> txn1 = test.beginTxn();
        Transaction<String> txn2 = test.beginTxn();

        txn1.writeEvent("1");
        txn1.flush();
        txn2.writeEvent("2");
        txn2.flush();

        Map<Double, Double> map = new HashMap<>();
        map.put(0.0, 1.0 / 3.0);
        map.put(1.0 / 3.0, 2.0 / 3.0);
        map.put(2.0 / 3.0, 1.0);
        Stream stream = new StreamImpl("test", "test");
        ScheduledExecutorService executor = Executors.newScheduledThreadPool(10);
        controller.scaleStream(stream, Arrays.asList(0L, 1L, 2L), map, executor);

        txn1.commit();
        while (txn1.checkStatus().equals(Transaction.Status.COMMITTING)) ;
        System.err.println(" transaction1 committed");
        // above should have rolled over
        txn2.commit();
        while (txn1.checkStatus().equals(Transaction.Status.COMMITTING)) ;
        System.err.println(" transaction1 committed");
        // above should have rolled over second time
        waitForIt.join();
        System.exit(0);
    }

}
```
